File: frontend/src/lib/rbac.ts
// Mirror of backend/app/rbac.py — keep in sync.
export type Role = "fleet_manager" | "dispatcher" | "safety_officer" | "financial_analyst";
export type Resource =
  | "dashboard" | "fleet" | "drivers" | "trips"
  | "maintenance" | "fuel_expenses" | "analytics" | "settings";
export type Access = "full" | "view" | null;

export const MATRIX: Record<Role, Record<Resource, Access>> = {
  fleet_manager: {
    dashboard: "full", fleet: "full", drivers: "full", trips: "view",
    maintenance: "full", fuel_expenses: "view", analytics: "full", settings: "full",
  },
  dispatcher: {
    dashboard: "full", fleet: "view", drivers: "view", trips: "full",
    maintenance: null, fuel_expenses: null, analytics: null, settings: null,
  },
  safety_officer: {
    dashboard: "full", fleet: null, drivers: "full", trips: "view",
    maintenance: null, fuel_expenses: null, analytics: null, settings: null,
  },
  financial_analyst: {
    dashboard: "full", fleet: "view", drivers: null, trips: "view",
    maintenance: "view", fuel_expenses: "full", analytics: "full", settings: null,
  },
};

export function can(role: Role | undefined, resource: Resource, write = false): boolean {
  if (!role) return false;
  const access = MATRIX[role]?.[resource] ?? null;
  return write ? access === "full" : access !== null;
}

export const ROLE_LABELS: Record<Role, string> = {
  fleet_manager: "Fleet Manager",
  dispatcher: "Dispatcher",
  safety_officer: "Safety Officer",
  financial_analyst: "Financial Analyst",
};

File: backend/app/rbac.py
"""RBAC matrix (mirrored in frontend/src/lib/rbac.ts — keep in sync).

Values: "full" (read+write), "view" (read-only), None (no access).
Resources map to routers; "dashboard" is readable by everyone.
"""

from fastapi import Depends, HTTPException

from .auth import get_current_user
from .models import User

MATRIX: dict[str, dict[str, str | None]] = {
    "fleet_manager": {
        "dashboard": "full",
        "fleet": "full",
        "drivers": "full",
        "trips": "view",
        "maintenance": "full",
        "fuel_expenses": "view",
        "analytics": "full",
        "settings": "full",
    },
    "dispatcher": {
        "dashboard": "full",
        "fleet": "view",
        "drivers": "view",
        "trips": "full",
        "maintenance": None,
        "fuel_expenses": None,
        "analytics": None,
        "settings": None,
    },
    "safety_officer": {
        "dashboard": "full",
        "fleet": None,
        "drivers": "full",
        "trips": "view",
        "maintenance": None,
        "fuel_expenses": None,
        "analytics": None,
        "settings": None,
    },
    "financial_analyst": {
        "dashboard": "full",
        "fleet": "view",
        "drivers": None,
        "trips": "view",
        "maintenance": "view",
        "fuel_expenses": "full",
        "analytics": "full",
        "settings": None,
    },
}


def require(resource: str, write: bool = False):
    def dep(user: User = Depends(get_current_user)) -> User:
        access = MATRIX.get(user.role, {}).get(resource)
        if access is None or (write and access != "full"):
            raise HTTPException(403, f"Role '{user.role}' cannot {'modify' if write else 'access'} {resource}")
        return user

    return dep
